src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,449 @@
+{
+	/**
+	 * @brief The value to be returned.
+	 * This is implemented as a separate variable to allow for much easier debugging, as
+	 * opposed to just instantly returning the value read from the input lines.
+	 */
+	uint8_t value = 0;
+
+	// Note: Input line 4 covers the keyboard circuit, which is not implemented here.
+	// When the keyboard state is read, the default value of 0 will be returned.
+	if(m_key_switch_scan_driver_select == 0)
+	{
+		// Input line 0 covers the 'main' front-panel switches.
+		value = m_key_switch_input_driver_line_0->read();
+	} else if(m_key_switch_scan_driver_select == 1)
+	{
+		// Input line 1 covers the numeric front-panel switches 1 through 8.
+		value = m_key_switch_input_driver_line_1->read();
+	} else if(m_key_switch_scan_driver_select == 2)
+	{
+		// Input line 2 covers the numeric front-panel switches 9 though 16.
+		value = m_key_switch_input_driver_line_2->read();
+	} else if(m_key_switch_scan_driver_select == 3)
+	{
+		// Input line 3 covers the numeric front-panel switches 17 though 20.
+		// The Portamento, and Sustain pedals are mapped to bits 6, and 7 respectively.
+		value = m_key_switch_input_driver_line_3->read();
+	}
+
+	return value;
+}
+
+
+/**
+ * yamaha_dx9_state::unmapped_w
+ */
+void yamaha_dx9_state::unmapped_w(offs_t address, uint8_t data) {}
+
+
+/**
+ * yamaha_dx9_state::ops_w
+ */
+void yamaha_dx9_state::ops_w(offs_t offset, uint8_t data)
+{
+	// printf("OPS: %02X=%02X\n", offset, data);
+}
+
+
+/**
+ * yamaha_dx9_state::egs_w
+ */
+void yamaha_dx9_state::egs_w(offs_t offset, uint8_t data)
+{
+	// printf("EGS: %02X=%02X\n", offset, data);
+}
+
+
+/**
+ * yamaha_dx9_state::led_w
+ */
+void yamaha_dx9_state::led_w(offs_t offset, uint8_t data)
+{
+	// Since the memory mapped region that calls this function is only two byts in
+	// size, the led number is the least-significant bit of the offset.
+	// The DX9's LEDs are wired so that a high input line disables the segment, so
+	// get the one's complement of the data.
+	m_leds[offset & 1] = (~data) & 0xFF;
+}
+
+
+/**
+ * yamaha_dx9_state::p1_r
+ */
+uint8_t yamaha_dx9_state::p1_r(offs_t offset)
+{
+	// The ADC EOC line is wired to bit 4, as well as the Cassette Interface input, which
+	// is wired to bit 7. This is currently not fully implemented.
+	return m_adc->eoc_r() << 4;
+}
+
+
+/**
+ * yamaha_dx9_state::p1_w
+ */
+void yamaha_dx9_state::p1_w(offs_t offset, uint8_t data)
+{
+	// The low-nibble is written by the firmware to select the key/switch driver input line.
+	m_key_switch_scan_driver_select = data & 0xF;
+
+	// The cassette interface remote port polarity is set by bit 5.
+	m_cassette_interface_remote_polarity = data & 0x20;
+
+	// The cassette interface output polarity is set by bit 6.
+	m_cassette_interface_output_polarity = data & 0x40;
+}
+
+
+static INPUT_PORTS_START(dx9)
+	PORT_START("KEY_SWITCH_INPUT_DRIVER_LINE_0")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_BUTTON1)    // Front-panel button Yes/Up.
```

Comment:
Use `IPT_KEYBOARD` and `PORT_NAME` to name the keys.

Reply:
I used `IPT_UNKNOWN` over `IPT_KEYBOARD`, as the latter seems to capture keyboard input.